lib/galaxy_test/api/test_history_contents.py

```diff
@@ -1266,6 +1266,38 @@ def test_bulk_datatype_change(self):
                 assert item["data_type"] == "galaxy.datatypes.tabular.Tabular"
                 assert "metadata_column_names" in item
 
+    def test_bulk_datatype_change_should_skip_set_metadata_on_deferred_data(self):
+        with self.dataset_populator.test_history() as history_id:
+            details = self.dataset_populator.create_deferred_hda(
+                history_id, "https://raw.githubusercontent.com/galaxyproject/galaxy/dev/test-data/1.bed", ext="bed"
+            )
+            assert details["state"] == "deferred"
+            assert details["extension"] == "bed"
+            assert details["data_type"] == "galaxy.datatypes.interval.Bed"
+            assert details["metadata_columns"]
+
+            new_datatype = "txt"
+            payload = {
+                "operation": "change_datatype",
+                "params": {
+                    "type": "change_datatype",
+                    "datatype": new_datatype,
+                },
+            }
+            bulk_operation_result = self._apply_bulk_operation(history_id, payload)
+            self._assert_bulk_success(bulk_operation_result, expected_success_count=1)
+
+            # Wait for celery tasks to finish
+            time.sleep(2)  # I don't like this at all, is there another way to wait for celery here?
```

Comment:
Is there a better way to wait for celery to finish here? Unfortunately waiting for a history change doesn't help.

Reply:
You're not running a celery task, so ... I don't think you need to wait ? Just assert that the dataset is not in `setting_metadata` should be suffcient.

Reply:
So my first comment is nonsense, but changing the datatype without setting metadata doesn't have to happen in a celery task, there's very little I/O or CPU involved, I think it's fine to do that during the request for now.

Reply:
Is c58e3f4 more or less what you had in mind?

Reply:
Yes, perfect.